src/library/browse/foldertreemodel.h

```diff
@@ -16,13 +35,41 @@ class TreeItem;
 
 class FolderTreeModel : public TreeItemModel {
     Q_OBJECT
+
+    using FolderQueue = QQueue<std::pair<QModelIndex, QString>>;
+    // FIXME: this should be a Trie (prefix tree)
+    using Cache = std::unordered_map<QString, bool>;
+
   public:
-    FolderTreeModel(QObject *parent = 0);
+    using TreeItemList = QList<TreeItem*>*;
+
+    FolderTreeModel(QObject* parent = 0);
     virtual ~FolderTreeModel();
     virtual bool hasChildren(const QModelIndex& parent = QModelIndex()) const;
-    bool directoryHasChildren(const QString& path) const;
+    void processFolder(const QModelIndex& parent, const QString& path) const;
+
+  private:
+    void directoryHasChildren(const QString& path);
+
+  private slots:
+    void dirModified(const QString& str);
+    void addChildren(const QModelIndex& parent, TreeItemList children);
+    void onHasSubDirectory(const QString& path);
+
+  signals:
+    void newChildren(const QModelIndex& parent, TreeItemList hildren);
+    void hasSubDirectory(const QString& path);
 
   private:
     // Used for memoizing the results of directoryHasChildren
-    mutable QHash<QString, bool> m_directoryCache;
+    Cache m_directoryCache;
+    mutable MReadWriteLock m_cacheLock;
+    QFileSystemWatcher m_fsWatcher;
+    QThreadPool m_pool;
+    mutable FolderQueue m_folderQueue;
+    mutable std::mutex m_queueLock;
+    std::atomic<bool> m_isRunning;
+    QFuture<void> m_folderProcess;
```

Comment:
QFuture handles threads, not processes. How about  `m_folderFuture` or  `m_folderThread`